### PR TITLE
Add an option to require ignore comments have error codes

### PR DESCRIPTION
**Pull request by PeterJCLaw**
### Description

Fixes https://github.com/python/mypy/issues/11509

> **Pitch** `# type:ignore` comments risk suppressing an unintended error, which is why it's preferred to specify the exact error code you're suppressing

This introduces a new error code `ignore-without-code` which achieves that. ~I've gone with naming this `warn-` rather than `disallow-` since the ignores are still obeyed, just alerted about (and also as the most closely related options are also `warn-` style).~ This is disabled by default but can be controlled via the `--enable-error-code` machinery.

This new option is aware of `--warn-unused-ignores` (since they're likely to be used together) such that we don't emit warnings about a lack of error codes if the ignore is unused and that option is turned on.

~One thing I'm not sure about is what should happen for ignore comments that apply to a whole file. I've raised this on that issue, so discussion should probably happen there.~ Currently the new option obeys that instruction to ignore the file.

### Test Plan

Some manual testing, plus automated tests added.

### Code Review

~I'm fairly happy with the code approach here, though I'm not at all sure that I've put the command line option in the right places. I also suspect I've missed some places that would need updating in the docs. Guidance around the proper places to put the docs/CLI arg/etc. is most welcome.~ (I believe I've now caught everywhere that this needs to go)

### CODE REVIEW

**Comment by JelleZijlstra** [comment]
The mypy-primer output here is unexpected, right? The new error is off by default so it shouldn't trigger on existing code.

**Comment by PeterJCLaw** [comment]
> The mypy-primer output here is unexpected, right? The new error is off by default so it shouldn't trigger on existing code.

Indeed, I'm not sure what's causing that. When I run `mypy` from this branch by default I'm not seeing the option take effect.

Separately, I have noticed that if I run `mypy` from this branch initially without the new option and then run it with the new option the errors aren't found when they should be:
```$ cat foo.py
foo  # type: ignore
$ mypy foo.py
Success: no issues found in 1 source file
$ mypy foo.py --disallow-ignore-without-code
Success: no issues found in 1 source file
```
Changing the test file seems to get the error to show up, so I'm wondering if there's some caching that I've missed?

**Comment by hauntsaninja** [comment]
I think you might have automatically included it in `--strict` (see the strict_group stuff)

**Comment by PeterJCLaw** [comment]
> I think you might have automatically included it in `--strict` (see the strict_group stuff)

Ahhh, yeah. That was actually intentional and I'd forgotten I'd done. I'm not sure whether this qualifies for `--strict` (or how that gets decided), but it's something I'd be interested in including in the strict mode.

Perhaps it would be better to introduce this flag first and then later move it into `--strict`?
(If `--warn-unused-ignores` isn't in `--strict` then adding it at the same time might also make sense)

**Comment by hauntsaninja** [comment]
Yeah, I think we should introduce this first and only later consider moving it into `--strict`, since this has a pretty large impact on users of `--strict`. I'd also want to make sure that the error codes mypy provides for common errors are useful before doing so to reduce future breakage; I think currently we have a lot of error code "misc".

My approximate mental model for `--strict` is "you shouldn't ever have type errors without being explicit somewhere". So if today was the first day of mypy I'd be happy to include it, in our current world I lean towards avoiding breaking unsuspecting users since it's just an adjustment of something that's already explicitly opting in to type unsafety.

`--warn-unused-ignores` is already part of `--strict`, but `--show-error-codes` is not (but should be, if this is incorporated into `--strict`). Speaking of which, if possible, it would be really nice to show which error code(s) to add in your error message.

**Comment by PeterJCLaw** [comment]
> Yeah, I think we should introduce this first and only later consider moving it into `--strict`, since this has a pretty large impact on users of `--strict`. I'd also want to make sure that the error codes mypy provides for common errors are useful before doing so to reduce future breakage; I think currently we have a lot of error code "misc".

Ah, fair. I've personally found that while a lot of things were `misc`, most things now aren't (:+1: to whoever was pushing that a while back). I can definitely see the desire to get `misc` even lower before requiring codes though :+1:

> ... if possible, it would be really nice to show which error code(s) to add in your error message.

Good point, I think it should be easy to add that.

**Comment by PeterJCLaw** [comment]
> Separately, I have noticed that if I run `mypy` from this branch initially without the new option and then run it with the new option the errors aren't found when they should be

I believe I've fixed this in https://github.com/python/mypy/pull/11633/commits/c811aa7fb13ef49c9f4677e6d925d10ca8e44e74.

**Comment by JukkaL** [comment]
I think this would be better enabled via ` --enable-error-code <something>`, instead of adding a new command-line option. The motivation is to avoid continuously increasing the size and complexity of the command-line interface. 

(Some of the existing flags would work better using `--enable-error-code`, but they were added before `--enable-error-code` was available and maintaining backward compatibility is useful.)

**Comment by PeterJCLaw** [comment]
@JukkaL thanks, I've changed over to this being an error code (still disabled by default).

**Comment by PeterJCLaw** [comment]
@JukkaL just wondering if there's anything more I need to do here before this can be reviewed? (Apologies if you're not the right person to ask)

**Comment by PeterJCLaw** [comment]
@JelleZijlstra just wondering if there's anything more I need to do here before this can be reviewed? (Apologies if you're not the right person to ask)

(I realise there's a merge conflict here, though it looks fairly easy to resolve -- do say if that needs fixing before this can be reviewed)

**Comment by JelleZijlstra** [comment]
I can handle getting this through, since Jukka and Shantanu already commented here without objections to the overall idea. If you fix the conflict, I'll do another round of review, then merge if everything looks good.

**Comment by PeterJCLaw** [comment]
Thanks! I'll fix that conflict now :+1:

**Comment by JelleZijlstra** [comment]
Looks like a test needs to be updated.

File: docs/source/error_code_list2.rst [hunk 1]
```diff
@@ -255,3 +255,31 @@ no error in practice. In such case, it might be prudent to annotate ``items: Seq
 This is similar in concept to ensuring that an expression's type implements an expected interface (e.g. ``Sized``),
 except that attempting to invoke an undefined method (e.g. ``__len__``) results in an error,
 while attempting to evaluate an object in boolean context without a concrete implementation results in a truthy value.
+
+
+Check that ``# type: ignore`` include an error code
```
Comment (JelleZijlstra) [inline docs/source/error_code_list2.rst:260]:
```suggestion
Check that ``# type: ignore`` include an error code [ignore-without-code]
```

Include the error code in the section title, as with the previous section

File: docs/source/error_code_list2.rst [hunk 2]
```diff
@@ -255,3 +255,31 @@ no error in practice. In such case, it might be prudent to annotate ``items: Seq
...
+
+    # mypy: enable-error-code ignore-without-code
+
+    class Foo:
+        def __init__(self, name: str) -> None:
+            self.name = name
+
+    f = Foo('foo')
+
+    # This line has a typo that mypy can't help with as both the expected 'assignment' and 'attr-defined' are silenced
```
Comment (JelleZijlstra) [inline docs/source/error_code_list2.rst:279]:
```suggestion
    # This line has a typo that mypy can't help with as both the expected error 'assignment' and the unexpected error 'attr-defined' are silenced
```

This was a bit hard to read. Probably should also split it into multiple lines.

**Review by JelleZijlstra — CHANGES_REQUESTED** [review]
Some minor feedback on the docs

**Comment by NeilGirdhar** [comment]
Just curious, what happens if you have a `# type: ignore[pyright]` or some other code that doesn't exist?

File: docs/source/error_code_list2.rst (hunk 1 above)
Comment (PeterJCLaw) [reply docs/source/error_code_list2.rst]: f9e60f9b36eda3443a69481bbffed9e293c87833 adds this

File: docs/source/error_code_list2.rst (hunk 2 above)
Comment (PeterJCLaw) [reply docs/source/error_code_list2.rst]: 7798cbd902b3c6c9a781eba1e5ac857ef5dbe355 hopefully helps here?

**Review by JelleZijlstra — APPROVED** [review]
Thank you!

**Comment by PeterJCLaw** [comment]
One thing I have spotted here is that it's now possible to have several errors from the same problem, as shown by the `testErrorCodeMissingDoesntTrampleUnusedIgnoresWarning` test:
``` python
# flags: --enable-error-code ignore-without-code --warn-unused-ignores
z # type: ignore[ignore-without-code]
```
generates three messages:
- Error: Unused "type: ignore" comment
- Error: Name "z" is not defined  [name-defined]
- Note: Error code "name-defined" not covered by "type: ignore" comment

While I'm not sure if this is something created by this PR, is this something we want to address here?

**Comment by JelleZijlstra** [comment]
> While I'm not sure if this is something created by this PR, is this something we want to address here?

I'd rather cover that in a separate PR. Maybe we should suppress the `Unused "type: ignore" comment` error?

File: docs/source/error_code_list2.rst [hunk 3]
```diff
@@ -255,3 +255,34 @@ no error in practice. In such case, it might be prudent to annotate ``items: Seq
 This is similar in concept to ensuring that an expression's type implements an expected interface (e.g. ``Sized``),
 except that attempting to invoke an undefined method (e.g. ``__len__``) results in an error,
 while attempting to evaluate an object in boolean context without a concrete implementation results in a truthy value.
+
+
+Check that ``# type: ignore`` include an error code [ignore-without-code]
+-------------------------------------------------------------------------
+
+Warn when an ``# type: ignore`` comment does not specify any error codes.
+This clarifies the intent of the ignore as well as ensuring that only the
```
Comment (JelleZijlstra) [inline docs/source/error_code_list2.rst:264]:
```suggestion
Warn when a ``# type: ignore`` comment does not specify any error codes.
This clarifies the intent of the ignore and ensures that only the
```
Sorry, one more change

**Comment by DanielNoord** [comment]
Just want to express my gratitude to @PeterJCLaw for writing and @JelleZijlstra for reviewing this PR. We have been waiting patiently for this function, nice to see it got merged 🎉 Thanks guys!

**Pull request merged**